src/core/server/opensearch_data/client/data_source_client.ts

```diff
@@ -0,0 +1,130 @@
+   * create a new client instance to be able to interact with OpenSearch API.
+   */
+  close: () => Promise<void>;
+}
+
+// TODO: tmp
+interface DataSourceInfo {
+  endpoint: string;
+  credentialId: string;
+}
+
+interface CredentialInfo {
+  username: string;
+  password: string;
+}
+
+export class DataSourceClient implements ICustomDataSourceClient {
+  public dataSourceClientsPool: Map<string, Client>;
+  private savedObjectClient: SavedObjectsClientContract;
+  private isDataSourceFeautureEnabled = true;
+  private isClosed = false;
+
+  constructor(
+    private readonly config: OpenSearchClientConfig,
+    savedObjectClient: SavedObjectsClientContract,
+    logger: Logger
+  ) {
+    // init pool as empty
+    this.dataSourceClientsPool = new Map<string, Client>();
+    this.savedObjectClient = savedObjectClient;
+    // TODO: 1.read config and determine isDataSourceEnabled Flag
+    // 2. throw error if isDataSourceEnabled == false, while API is called
+  }
+  async asDataSource(dataSourceId: string) {
+    const { endpoint, credentialId } = await this.getDataSourceInfo(dataSourceId);
+    const { username, password } = await this.getCredentialInfo(credentialId);
+    // 2. build/find client and return
+    let dataSourceClient = this.dataSourceClientsPool.get(dataSourceId);
+    if (!dataSourceClient) {
+      // TODO: make use of existing default clientConfig to build client
+      dataSourceClient = new Client({
+        node: endpoint,
+        auth: {
+          username,
+          password,
+        },
+      });
+      // update pool
+      this.dataSourceClientsPool.set(dataSourceId, dataSourceClient);
+    }
+    return dataSourceClient;
+  }
+
+  private async getDataSourceInfo(dataSourceId: string): Promise<DataSourceInfo> {
+    // 1. fetch meta info of data source using saved_object client
+    const dataSource = await this.savedObjectClient.get('data-source', dataSourceId);
+    // 2. TODO: parse to DataSource object, need update once dataSource type is in place
+    const dataSourceObj = dataSource!.attributes as any;
+    const endpoint = dataSourceObj.endpoint;
+    const credentialId = dataSource!.references[0].id;
+
+    return { endpoint, credentialId };
+  }
+
+  private async getCredentialInfo(credentialId: string): Promise<CredentialInfo> {
+    /**
+     * TODO:
+     * credential manager will provide "decrypt(authId: string)" to return auth
+     * Example code: cosnt {username, password} = credentialManager.decrpt(dataSourceObj.authId)
+     */
+    const credential = await this.savedObjectClient.get('credential', credentialId);
+    const credentialObj = credential!.attributes as any;
+    const { user_name: username, password: encryptedPassword } = credentialObj.credential_material;
+
+    const password = await CryptoCli.getInstance().decrypt(
+      Buffer.from(encryptedPassword, 'base64')
+    );
+    return { username, password };
+  }
+
+  // close anything in pool
+  public async close() {
```

Comment:
please check the opensearch js client, see if it can keep the connection until being explicitly closed

Reply:
https://github.com/elastic/elasticsearch-js/blob/4def742e4774dba44f31b8e1701345a8c29870ee/lib/Connection.js#L59-L63
`keep-alive` is default to true in `ClientOptions`